### DISCUSSION TITLE
Poincaré Classification Layer never converges

### DISCUSSION THREAD

**Discussion by tsteffek** [geoopt/geoopt#170] [Q&A]
Hi,

I'm trying to use geoopt to create a classification layer on top of a classic, Euclidean model (a BERT model). I'm using a default Poincaré Ball (c=1 / curvature at -1, non-learnable). The Euclidian version has a 2-layer FF classification net, and my first simple approach was to replace that with the MobiusLinear from your examples. So basically:
```python
x = bert(x)
x = ball.expmap0(x)
x = MobiusLinear(x) // to projection dimension
x = MobiusLinear(x) // to output dimension
```
Followed by PyTorch's [binary_cross_entropy_with_logits](https://pytorch.org/docs/stable/nn.functional.html#binary-cross-entropy-with-logits) and the `RiemannianAdam`.

Replacing the last layer with `Distance2PoincareHyperplanes` from the other example works and converges, but the version with `MobiusLinear` quickly stops learning anything and the loss refuses to change afterwards, even when overfitting a single batch with a learning rate of `0.1`. I can't think of a reason why this shouldn't work or why it should behave this strangely. Is there something I'm missing?

**Comment by SomeoneSerge** [comment]
Hi, sorry for guessing at random, but are you sure you didn't mean `expmap` in `x = ball.logmap0(x)`?

**Reply by tsteffek** [reply]
When rereading it just now I thought the same. And yes, I indeed meant `expmap0`. Corrected it. Thanks for the catch.

**Comment by ferrine** [comment]
There are no logits after linear layer so the loss doesn't make sense. The output is on the hyperbolic plane. You should use distances to hyperplanes

**Reply by tsteffek** [reply]
That confirms what I've discovered, but leaves the question why. After all, an output on the hyperbolic plane is still "just a vector" with the constraint that everything's within `[-1,1]` for the Poincaré Ball, is it not? `mobius_matvec` doesn't seem to be that much non-linearity to explain either.

Maybe I'm lacking in Riemannian geometry fundamentals to understand. If you happen to know good resources, I'd gladly take them.

**Reply by SomeoneSerge** [reply]
Ah, well, your "logits" must be within `[-1, 1]` then, imagine putting that into sigmoid

**Reply by tsteffek** [reply]
Well they will be squashed to ~`[0.3, 0.6]`. I considered that that may be too small to reasonably train, so I also tried ReLu instead of sigmoid. That way I'd only lose the `[-1, 0]`, but to no avail.
